core/corehttp/routing.go

```diff
@@ -0,0 +1,111 @@
+package corehttp
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/ipfs/boxo/routing/http/server"
+	"github.com/ipfs/boxo/routing/http/types"
+	"github.com/ipfs/boxo/routing/http/types/iter"
+	cid "github.com/ipfs/go-cid"
+	core "github.com/ipfs/kubo/core"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/multiformats/go-multiaddr"
+)
+
+const (
+	streamingProvidersCount    = 0
+	nonStreamingProvidersCount = 20
+)
+
+func RoutingOption() ServeOption {
+	return func(n *core.IpfsNode, _ net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
+		handler := server.Handler(&contentRouter{n})
+		mux.Handle("/routing/v1/", handler)
+		return mux, nil
+	}
+}
+
+type contentRouter struct {
+	n *core.IpfsNode
+}
+
+func (r *contentRouter) FindProviders(ctx context.Context, key cid.Cid, streaming bool) (iter.ResultIter[types.ProviderResponse], error) {
+	ctx, cancel := context.WithCancel(ctx)
+	count := nonStreamingProvidersCount
+	if streaming {
+		count = streamingProvidersCount
+	}
+	ch := r.n.Routing.FindProvidersAsync(ctx, key, count)
+	return iter.ToResultIter[types.ProviderResponse](&peerChanIter{
+		ch:     ch,
+		cancel: cancel,
+	}), nil
+}
+
+func (r *contentRouter) Provide(ctx context.Context, req *server.WriteProvideRequest) (types.ProviderResponse, error) {
+	// Kubo /routing/v1 endpoint does not support write operations.
+	return nil, routing.ErrNotSupported
+}
+
+func (r *contentRouter) ProvideBitswap(ctx context.Context, req *server.BitswapWriteProvideRequest) (time.Duration, error) {
+	// Kubo /routing/v1 endpoint does not support write operations.
+	return 0, routing.ErrNotSupported
+}
+
+type peerChanIter struct {
+	ch     <-chan peer.AddrInfo
+	cancel context.CancelFunc
+	next   *peer.AddrInfo
+}
+
+func (it *peerChanIter) Next() bool {
+	addr, ok := <-it.ch
+	if ok {
+		it.next = &addr
+		return true
+	} else {
+		it.next = nil
+		return false
+	}
+}
+
+func (it *peerChanIter) Val() types.ProviderResponse {
+	if it.next == nil {
+		return nil
+	}
+
+	// We don't know what type of protocol this peer provides. It is likely Bitswap
+	// but it might not be. Therefore, we set an unknown protocol with an unknown schema.
+	rec := &providerRecord{
+		Protocol: "transport-unknown",
+		Schema:   "unknown",
+		ID:       it.next.ID,
+		Addrs:    it.next.Addrs,
+	}
```

Comment:
On IPFS DHT, does it currently provide records with any transport protocol other than `BitSwap`?

On `unknown` transport, this seem to [exist implicitly](https://github.com/ipfs/boxo/blob/main/routing/http/types/provider_unknown.go) where if a client doesn't know the transport it passes it on as is? In a case where provider record is constructed, like above, an explicit `unknown` seems reasonable. 

@lidel  WDYT?

Reply:
Currently, it is only bitswap, but that is the problem: we are hardcoding it everywhere making it more and more difficult for alternative to be integrated in the future (part of the problems I've flagged in https://github.com/ipfs/specs/issues/377).

Going with `unknown` when we lack information is better than assuming bitswap. Client can decide to try providers with known protocols, and if those are not enough, connect to `unknown` ones  to see if they speak compatible transport.

I think its good enough, but wonder if we can make something better?


### Would "peer" schema be better than "unknown" transport?

We have other gap in the `/routing/v1`: peer routing (which we need to add for some types of light clients, at least to have an answer for DHT proxied/cached over HTTP).

Maybe we can avoid inventing fake "unknown" transport 
and instead create a `Schema` for `peer` that could in the future be reused in `/routing/v1/peers` ?

```json
{
 "Schema":   "peer",
 "ID":       "12FooPeerID",
 "Addrs":    [],
 "Protocols": ["transport-bitswap", "transport-http-gateway"], // optional, if we know
}
```

Proxied DHT results without known protocols would look like:

```json
{
 "Schema":   "peer",
 "ID":       "12FooPeerID",
 "Addrs":    ["/ip4/1.2.3.4/udp/4001/quic-v1/p2p/12PeerID2"],
 "Protocols": [], // empty means we dont know, client needs to connect and ise `identify` protocol to find out, or use a separate peer routing  call
}
```


@hacdias @masih  thoughts?

Reply:
@masih we get Peer IDs and, sometimes, addresses. That's all.

Reply:
Right so i think there is a bigger discussion here which is about expanding the scope of information returned by DHT provider records, which clearly goes beyond the scope of this PR.

Reply:
> The fact today is that all records returned by the DHT are using bitswap. Using unknown would make the protocol less useful but sure it is more abstract.

FYI that this isn't technically true. The IPFS Public DHT's provider records are used for things beyond rendezvous for Bitswap. As a simple example, pubsub applications (including IPNS over PubSub) advertise their interest in a topic as a provider record and Bitswap isn't involved at all.

So as pointed out above the information being conveyed is basically "a libp2p node has expressed affinity to this data, you can negotiate with them (or guess/know ahead of time) what protocols to use to interact with them.

Reply:
Ok, @hacdias  my understanding of next steps here (i'm OoO Wed):

- add a separate `peerRecord` struct  that follows `"Schema": "peer"`  from https://github.com/ipfs/kubo/pull/9877/files#r1195164693 
-  add `Schema: peer` to `/routing/v1/providers` under  [Known Transfer Protocols](https://specs.ipfs.tech/routing/http-routing-v1/#known-transfer-protocols) section, rename the section to `Known Schemas` 
   - could be new IPIP, or as part of https://github.com/ipfs/specs/pull/410 with rationale that we need it for streaming DHT proxy responses
  - once we have "unknown" case solved at spec level, we can ensure both Kubo and DHT proxying at  cid.contact will stop using undocumented `Schema: unknown` and switch to `Schema: peer`

Reply:
@lidel from your last point, it sounds like I should keep this PR as-is with the `unknown` schema, work on the specs, and only then circle back? But then you also ask to add a new struct to this PR on the first point, so I will want for when you come back. I will work on updating the IPIP.

Reply:
As per my 1:1 with @lidel, we will be blocking this PR until we introduce a new IPIP with the `Schema: peer`. Then, we will update Boxo and circle back to this PR.

Reply:
> The current spec is very specific on that: the provided addresses should work for the specified protocol.

Yes, that is why we need to clean this up, starting with  `peer` schema, without this often contradictory requirement.

For example, `Schema: bitswap`, which "transport-bitswap" is it? 1.0.0, 1.1.0, 1.2.0? :upside_down_face:  What if there is provider that speaks 2.0.0 and client only speaks 1.0.0? To make sense, that schema would have to have an additional  `Versions: [1.0.0, 1.1.0, 1.2.0]` field.  

Currently `Schema: bitswap` it not future-proof, and will probably end up being not used in the long term because sooner or later we will have providers that speak both bitswap and [HTTP](https://specs.ipfs.tech/http-gateways/trustless-gateway/) or graphsync and HTTP, and the HTTP endpoint will only be available on addrs with  `/http`, so in practice the majority of  providers will be returned  `peer` with multiple addrs and optional list of supported protocols to avoid returning the same peer multiple times or using `"unknown" schema.